### ISSUE TITLE
Connections not disposed: Max connections exception

### ISSUE THREAD

**Issue by dapalmi** [JasperFx/marten#1215]
I am using the DocumentStore in an ASP.NET Core 2.2 application.
Here the definition in the Startup.cs:

```
            var store = DocumentStore.For(_ =>
            {
                _.Connection(Configuration.GetConnectionString("PostgresDb"));
                _.DatabaseSchemaName = "search";
                var serializer = new JsonNetSerializer();
                serializer.Customize(c => c.ContractResolver = new ResolvePrivateSetters());
                _.Serializer(serializer);
                _.Events.UseAggregatorLookup(AggregationLookupStrategy.UsePrivateApply);
                _.Schema.Include<SearchMartenRegistry>();
            });
            services.AddSingleton<IDataAccess>(new DataAccess(store));
```

Then from the API controller I call this method in the `DataAccess` class:

```
        public IQueryable<Search> Search(string text)
        {
            using (var session = store.QuerySession())
            {
                return session.Query<Search>().Where(s => s.Name.Contains(text));
            }
        }
```

And in the controller a few extra things are executed on the `IQueryable` object (sort, skip, take, etc.).

If I understand the documentation right, the connection is established with the session and likewise disposed with it.

But the DB connections don't seem to be disposed, because after 100 time (default max connection in Postgres db) I get following error:
**The connection pool has been exhausted**

**Comment by mdissel** [comment]
your session is disposed when exiting the using clause.  Use

```
public List<Search> Search(string text)
        {
            using (var session = store.QuerySession())
            {
                return session.Query<Search>().Where(s => s.Name.Contains(text)).ToList();
            }
        }
```
or if you want to use locally (not in db) executed Linq: 

public IQueryable<Search> Search(string text)
        {
            using (var session = store.QuerySession())
            {
                return session.Query<Search>().Where(s => s.Name.Contains(text)).ToList().AsQueryable();
            }
        }

**Comment by dapalmi** [comment]
@mdissel That's what I thought too. But evidently it does not dispose the connection when exiting the using clause, because I get the max connections exception.

**Comment by mdissel** [comment]
Can you update the code with one of my examples and try again?

**Comment by oskardudycz** [comment]
@dapalmi ➕ for the @mdissel request.

Could you also give full usage scenario? So how are you using Search method?

**Comment by dapalmi** [comment]
@oskardudycz The search method is called from a api controller method:
```
        [HttpGet]
        public object Search(string text, DataSourceLoadOptions loadOptions)
        {
            var queryable = _searchDataAccess.Search(text);
            return DataSourceLoader.Load(queryable, loadOptions);
        }
```
The DataSourceLoader.Load() executes sort, skip, take, etc. on the queryable object.
@mdissel I can confirm that the addition of .ToList().ToQueryable() solved the problem.

So, if I return an IQueryable object the connection will not be disposed. Is that correct?

**Comment by mysticmind** [comment]
@dapalmi when you call `return session.Query<Search>().Where(s => s.Name.Contains(text))`, you are returning an instance of type `IMartenQueryable<T>` which holds a Postgres connection. So wrapping it in a using block does not magically dispose of the connection due to your pattern of usage.

As @mdissel indicated, by using `return session.Query<Search>().Where(s => s.Name.Contains(text)).ToList().AsQueryable()`, here `.ToList()` resulted in the execution of the query and the results returned. By calling `ToList().AsQueryable()`, you using `IQueryable` on the returned results.

In summary, your pattern of usage is causing the issue at hand.

**Comment by mysticmind** [comment]
Closing this. we can re-open if anyone thinks we need to add some documentation around usage patterns.

**Comment by kinl99** [comment]
@mysticmind Thanks for the explanation, we made the same mistake as @dapalmi. But we'd still like to use the repository pattern. Could you please give some advice on how to implement a repository where the lib of the repository is the only lib that holds a reference to marten but without sacrificing the generic query methods?

**Comment by mysticmind** [comment]
@kinl99 please take a look at https://github.com/oskardudycz/GoldenEye/tree/master/src/Core/Backend.Core.Marten, this is part of GoldenEye library, possibly you can use the same.

**Comment by kinl99** [comment]
Thanks @mysticmind!!!
Using dependency injection consequently, like in the example provided, not only resolved the problem but also removed quite some usings from the data layer.

**Comment by oskardudycz** [comment]
@mysticmind thank you for the recommending GoldenEye 👍 😄 
@kinl99 here you can find wrappers for the Marten https://github.com/oskardudycz/GoldenEye/tree/master/src/Core/Backend.Core.Marten/Context, they're called `DataContexts` but they're in fact following repository pattern.
I'd personally recommend to not do manual disposal of the DocumentSession but let the Dependency Injection container let manage it properly - so eg. use request scope registration for DocumentSession and Singleton for Document Store. Then you shouldn't have such issues.
For the future reference if that doesn't answer your question or we have

**Comment by egorpavlikhin** [comment]
Since Marten doesn't do connection pooling on its own it might be worth covering this in the documentation? The links above don't work anymore. Current example lives here https://github.com/oskardudycz/GoldenEye/blob/main/src/Marten/Marten/Registration/Registration.cs

**Comment by lamelyan** [comment]
@oskardudycz or @kinl99 Thank you for the idea of using DI to clean up the connections.  Could you point me to a sample of the code where you set the "request scope registration for DocumentSession and Singleton for Document Store."?
